// scaffdog/lib/template/ast.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
class Ident {
    constructor(name) {
        this.name = name;
    }
    toString() {
        return this.name;
    }
}
exports.Ident = Ident;
class Literal {
    constructor(value) {
        this.value = value;
    }
    toString() {
        return typeof this.value === 'string' ? `"${this.value}"` : `${this.value}`;
    }
}
exports.Literal = Literal;
class RawExpr {
    constructor(value) {
        this.value = value;
    }
    toString() {
        return this.value;
    }
}
exports.RawExpr = RawExpr;
class IdentExpr {
    constructor(ident) {
        this.ident = ident;
    }
    toString() {
        return this.ident.toString();
    }
}
exports.IdentExpr = IdentExpr;
class LiteralExpr {
    constructor(literal) {
        this.literal = literal;
    }
    toString() {
        return this.literal.toString();
    }
}
exports.LiteralExpr = LiteralExpr;
class CallExpr {
    constructor(name, args) {
        this.name = name;
        this.args = args;
    }
    toString() {
        const args = this.args.map((arg) => arg.toString()).join(', ');
        return `${this.name}(${args})`;
    }
}
exports.CallExpr = CallExpr;
class TagExpr {
    constructor(expressions) {
        this.expressions = expressions;
    }
    toString() {
        return `{{ ${this.expressions.map((expr) => expr.toString()).join('')} }}`;
    }
}
exports.TagExpr = TagExpr;
